### DISCUSSION TITLE
powerfox Poweropti als Power Meter zur Wechselrichter Steuerung einbinden

### DISCUSSION THREAD

**Discussion by olomberg** [hoylabs/OpenDTU-OnBattery#165] [Show and tell]
Hallo liebe Community.

In diesem Beitag möchte ich meine Erfahrung zur Einrichtung eines powerfox Poweropti als Power Meter (Stromzähler) mit euch teilen. Vielleicht kann es der ein oder andere gebrauchen.

In den "Einstellungen" die Option "Power Meter" auswählen.
"Aktiviere Stromzähler" auf "Ein" stellen.
"Stromzählertyp" auf "HTTP(S) + JSON" setzen.

"Individuelle HTTP requests pro Phase" auf "Aus" belassen/stellen, da der Poweropti nicht die Daten der einzelnen Phasen abgreift, sondern nur einen Gesamtwert liefert.

"URL" eingeben. Diese lautet "https://backend.powerfox.energy/api/2.0/my/DEVICE-ID/current". Dabei "DEVICE-ID" durch eure individuelle Poweropti Device-ID ersetzen. Ob diese dann richig ist, könnt ihr auch mittels Web Browser testen. Die URl dann einfach in den Browser eurer Wahl einfügen. Es sollte dann ein Fenster aufpoppen in welchem ihr die Anmeldedaten für den Poweropti eingeben könnt. Wenn ihr erfolgreich wart, solltet ihr die ein Zeile, änhliche der folgenden, im Browser sehen:

{"Outdated":false,"Watt":159,"Timestamp":1681191540,"A_Plus":18194947,"A_Minus":183262}

"Optional: HTTP request header - Key" ist "Authorization"

"Optional: HTTP request header - Wert" - Das ist jetzt der etwas kompliziertere Part, welcher aber nich allzu schwer ist. 

Aus der Doku der powerfox API : "Die Authentifizierung der Abfragen erfolgt via HTTP Basic Auth, d.h. E-Mail + Passwort des powerfox-Accounts im Base-64 Standard." Dazu braucht ihr einen Base 64 Encoder euerer Wahl, oder nutzt diesen hier: https://www.base64encode.org/
Dort gebt ihr die E-Mail Adresse eures powerfox Accounts, getrennt durch ":" (Doppel-Punkt), und euer Passwort ein.
Bsp.: ich@meine-mail.de:mysecretpassword

Diesen String nehmt ihr euch komplett in die Zwischenablage und fügt diesen in den Encoder ein. Anschließen nur noch auf Encode drücken. Somit wird aus "ich@meine-mail.de:mysecretpassword" der Base 64 encodete String "aWNoQG1laW5lLW1haWwuZGU6bXlzZWNyZXRwYXNzd29yZA=="

Diesem müssen wir jetzt noch die Authentifizierungsmethode voranstellen. Diese ist ja lt. API Doku "Basic". Somit erhalten wir:

"Basic aWNoQG1laW5lLW1haWwuZGU6bXlzZWNyZXRwYXNzd29yZA==" welche ihr bitte in "Optional: HTTP request header - Wert" einsetzt.

"Timeout" - Leider ist die API von powerfox manchmal nicht ganz so performant, so dass ihr den Timeout etwas hochsetzen solltet. Ich persönlich habe gute Erfahrungen mit einem Wert von 10 Sekunden gemacht. Das das Ganze in "ms" einzutragen ist, hier also den Wert auf "10000" setzen.

"JSON Pfad" - Jetzt müssen wir dem System noch mitgeben, welcher Pfad unseren Wert enthält. Wir ihr aus dem Test über den Browser entnehmen könnt ist dieses "Watt".

Jetzt könnt ihr Speichern und Testen. Dabei die Ruhe bewahren und ggf. Mehrfach auf Testen klicken. Wie gesagt, die API von powerfox ist etwas träge.

Ihr solltet dann auch nach dem Test eine Rückmeldung erhalten : "Success! Power: -79.00W"

Viel Erfolg bei der Einrichtung.

![image](https://user-images.githubusercontent.com/99411925/231071998-967cf028-7f12-4c48-8380-da9a9fc767c7.png)

**Comment by berni2288** [comment]
Hi @olomberg,

danke dass du deine Erfahrungen mit uns teilst!

> "Optional: HTTP request header - Wert" - Das ist jetzt der etwas kompliziertere Part, welcher aber nich allzu schwer ist.

Etwas doof dass das notwendig ist, aber ich vermute mal du hast es wegen dem doppelten "@" so gemacht, weil er sonst den Usernamen und Passwort nicht in der URL erkennt. Normalerweise sollte es ja auch im Format http://user:passwort@domain.tld möglich sein.
Vielleicht wäre eine extra Konfiguration für Basic Auth doch nicht so schlecht...

> "Timeout" - Leider ist die API von powerfox manchmal nicht ganz so performant, so dass ihr den Timeout etwas hochsetzen solltet. Ich persönlich habe gute Erfahrungen mit einem Wert von 10 Sekunden gemacht. Das das Ganze in "ms" einzutragen ist, hier also den Wert auf "10000" setzen.

Ein Timeout von 10 Sekunden ist schon extrem lang für den ESP32. In der Zeit kann der nichts anderes machen als auf die Antwort von Powerfox zu warten. Darunter wird wahrscheinlich die responsiveness ordentlich leiden, wenn es nicht sogar zu Neustarts kommt?
Wäre aber interessant was du damit für Erfahrungen machst :) Und wie lang Powerfox da mit macht bei so vielen Request :D

**Reply by olomberg** [reply]
Hi Berni.

Genau richtig. Es hängt an dem "@"-Zeichen, oder besser gesagt, dass der Username eine Mail-Adresse ist. Lt. RFC3986 ist das "@" ein "Reserved Character", genauso wie das ":"-Zeichen. Von daher der Umweg über den Header und den Value.

Der Timeout bei powerfox ist leider eine ganz traurige Geschichte. Aus meiner Sicht skaliert die Plattform nicht so richtig. Es gibt auch ein Limit von 128 concurrent calls auf die API.

Guter Hinweis bezüglich des "blockierens des ESP32. Lt. Console Messages habe ich keine Besonderheiten entdecken können. Vielleicht einfach mal mit 3000ms oder 5000ms starten und beobachten.
L
eider kommt man an den powerfox nur über die API an die Daten und nicht direkt per LAN. Habe noch einige von den Dingern im Bekanntenkreis mit Balkonkraftwerken, so dass ich nicht drum rum komme. Präferenz ist hier natürlich etwas lokales wie der Hichi IR.

Gruss
Oliver

**Reply by OlSteph** [reply]
Hallo zusammen,

vielleicht könnte hier der Poweropti auch lokal angesprochen werden, in meinem Homeassistant habe ich das über die configuration.yaml so gesteuert - leider mit POST und hier ist aber nur ein GET möglich.

Lokal kann ich sekündlich abfragen, damit wäre ein viel akkurateres Regeln möglich:

```python
rest:
  - resource: http://192.168.179.101/rpc
    method: POST
    payload: '{"id":1,"jsonrpc":"2.0","method":"getConfig","params":{"key":"latest_data"}}'
    scan_interval: 1
    headers:
      Content-Type: application/json
    sensor:
      - name: "poweropti_local"
        json_attributes:
          - "jsonrpc"
          - "id"
          - "result"
        value_template: >
            {% set json = value_json.result | base64_decode %}
            {% set Timestmp = (json | from_json())[0].t %}
            {% set zaehlernummer = (json | from_json())[0].m %}
            {% set A_Plus = (json | from_json())[0].d[0].v %}
            {% set A_Plus_HT = (json | from_json())[0].d[1].v %}
            {% set A_Plus_NT = (json | from_json())[0].d[2].v %}
            {% set A_Minus = (json | from_json())[0].d[3].v %}
            {% set timestmp2 = (json | from_json())[1].t %}
            {% set watt = (json | from_json())[1].d[0].v %}
            {{ '{"Watt":' + (watt | string) + ',"Timestamp":' + (Timestmp | string) + ',"A_Plus":' + (((A_Plus | float) / 1000) | string) + ',"A_Minus":' + (((A_Minus | float) / 1000) | string) + '}' }}

template:
  - sensor:
      - name: "Strom aktuellpoweropti lokal"
        unit_of_measurement: "W"
        device_class: "power"
        state_class: "measurement"
        state: >
            {{ (states('sensor.poweropti_local')|from_json).Watt }}
```

**Reply by Manos1966** [reply]
Danke.
uebrigens deine Zeile
{{ '{"Watt":' + (watt | string)  + ',"Timestamp":' + (Timestmp | string)  + ',"A_Plus":' + (((A_Plus | float) / 1000) | string) + ',"A_M>
ist unvollständig!
Korrekt sollte sein:
{{ '{"Watt":' + (watt | string)  + ',"Timestamp":' + (Timestmp | string)  + ',"A_Plus":' + (((A_Plus | float) / 1000) | string) + ',"A_Minus":' + (((A_Minus | float) / 1000) | string) + '}' }}

**Reply by OlSteph** [reply]
Ups, Danke. Habe ich so nicht gesehen - Hab's übernommen

**Comment by Manos1966** [comment]
Hallo @OlSteph 

ich mache bestimmt etwas falsch, bin aber zu dumm zu erkennen wo der Fehler liegt, hast du vielleicht eine Idee?

Als erstes,
Die Abfrage auf mein Chrome https://backend.powerfox.energy/api/2.0/my/main/current
nach Eingabe der Email Adresse und Passwort, gibt korrekt: 
{"Outdated":false,"Watt":883,"Timestamp":1691094573,"A_Plus":1894389,"A_Plus_HT":998,"A_Plus_NT":1893391,"A_Minus":2167} 
**_Ist es vielleicht weil es nicht meine DEVICE-ID eingegeben wurde? (obwohl diese hier funktioniert?)_**

Wenn ich aber versuche, bekomme ich nur **Bad HTTP code: 302**

**Reply by Manos1966** [reply]
OK, OK, "wer lesen kann ist im Vorteil" 👍 
Deine Beschreibung hat funktioniert! **Man muss die DEVICE-ID eingeben**. VIELEN DANK! 🥇 
Merkwürdigerweise, die Eingabe des Usernamens & Passworts hat nicht funktioniert. 
![Variations Correct](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/2b403f7a-1eb7-4011-9a6a-93493fea9e1f)

**Comment by Ikarus67** [comment]
Vielen Dank für die Beschreibung zur Einrichtung des Poweropti!
Allerdings funktioniert die Abfrage mit den neuen Programmversionen ab 2024 nicht mehr!? Bad HTTP Code: 404
Mit den Versionen vom Dezember lieferte die Abfrage noch verlässliche Werte.

Ist das schon aufgefallen, bzw. gibt es eine Lösung?

**Reply by Narcodoc** [reply]
Ist mir aufgefallen. Gibt es eine Lösung?

**Comment by OlSteph** [comment]
Hallo @Ikarus67 ,
ich habe persönlich eine permanente Lösung dafür (zumindest für mich) -> Tasmota basierter Hichi Stromzähler und meine Homeassistant Integration. Der Poweropti ist für mich nicht mehr nutzbar und auf Aussagen vom Support 
Auszug aus der Antwort zu meiner E-Mail Anfrage mit Stand 14.09.2023:

"Ein lokaler Zugriff auf die poweropti Daten ist aufgrund der Erhöhung unserer Sicherheitsstandards vorerst nicht mehr möglich. Es wird aber aktuell daran gearbeitet diesen lokalen Zugriff nach den Anforderungen unserer neuen Sicherheitsstandards wieder zu ermöglichen. Ein genauer Zeitpunkt der Umsetzung kann noch nicht abgeschätzt werden."

-> Darauf möchte/kann ich einfach nicht warten und daher ist der Poweropti bei mir raus geflogen. Regelung via Cloud viel zu langsam und unstabil, da häufig gestört (meine Erfahrung). Ob das aktuell wieder verfügbar ist könntest du beim Support dort in Erfahrung bringen:
support@powerfox.energy

_____________________
Edit: du beschreibst die Cloud-Schnittstelle und ich hatte die Lokale im Kopf, daher ist das Thema von mir hier verfehlt ;-(.
Dennoch kann ich das nicht (mehr) nachstellen, vielleicht hat noch wer anderes hier die Möglichkeit.

**Reply by OlSteph** [reply]
@Ikarus67 
Ach guck  -akuelle Version installiert und die Werte werden weniger oft aktualisiert ( bzw. angezeigt) 
Wahrscheinlich siehe "Breaking Changes"
https://github.com/helgeerbe/OpenDTU-OnBattery/releases/tag/2024.02.19.post1

**Comment by Billys67** [comment]
Hallo Zusammen, erstmal Danke an @olomberg für die Anleitung. Ich weiß nicht warum aber ich kriege es so nicht hin. Ich habe die letzte Version von openDTU im Einsatz (Extra Auswahlmenü für Basic Konfiguration)  Eine direkte Abfrage beim Poweropti in Browser liefert die Daten wie erwartet. Weiter geht es nicht. Egal ob mein Login Daten mit entcoder64 oder per direkte Eingabe in Basic Auswahlmenü eingebe, kommt immer die Fehlermeldung "Bad HTTP code: 404" . Liegt vielleicht an der neue version oder mach ich was falsch? Danke.

**Comment by Omega13x** [comment]
Habe mich auch ein Jahr mit dem Poweropti herumgeärgert und habe dann einfach Tasmota drauf geflasht. Läuft super und cloudfrei.
![image](https://github.com/user-attachments/assets/fb5b9c80-dfc0-4069-8c25-869baca74745)

**Reply by mi-hol** [reply]
@Omega13x 

> Habe mich auch ein Jahr mit dem Poweropti herumgeärgert und habe dann einfach Tasmota drauf geflasht

Damit ich es 100% richtig verstehe, hast Du die original Powerfox poweropti+  Hardware (PB202001) auf die Tasmota version mit SML geflashed?

Diese Hardware hat ja den Vorteil, dass die Stromversorgung direkt von eBZ und easymeter-Stromzählern kommt und die Installation damit "plug-and-play" ist.

**Comment by Omega13x** [comment]
Ich habe diesen Poweropti mit externem Netzteil.
![image](https://github.com/user-attachments/assets/e4536641-20b9-48f3-965a-506c728f2561)

An meinem zweiten Zähler habe ich mir dann, für ein drittel des Preises eines Poweropti, einen Bitshake Lesekopf montiert.

**Reply by thomluther** [reply]
@Omega13x, kannst du uns bitte eine kurze Anleitung geben wie du deinen Poweropti mit Tasmota SML geflashed hast damit er anschließend noch nutzbar war?
Bin auch am überlegen das zu machen weil lokale RPC Schnittstelle abgesichert wurde und Cloud nicht nutzbar ist für meine Zwecke. Habe noch einen Hichi Tasmota Lesekopf für meinen zweiten Zähler. Das SML Skript hätte ich also schon wenn der Poweropti mit Tasmota laufen würde.

**Comment by Opeloli** [comment]
Hallo,

ich möchte den o.g. Punkt gerne nochmal aufgreifen und erneut aufwärmen.

Ich stehe nämlich vor dem identischen Problem.
Der Poweropti von Powerfox sendet im Browser auf die Eingabe https://backend.powerfox.energy/api/2.0/my/main/current völlig problemlos und zuverlässig seine aktuellen Werte.

Bei der Eingabe der Daten in OpenDTU kommt immer die Fehlermeldung "Bad HTTP Code:404"

und das sowohl bei direkter Eingabe der Zugangs Daten mit "Authorization Type: Basic", "Username: login in Powerfox (=E-Mailadresse)", Password

als auch über den Optionalen Http: Request mit der Base64 Encodierung.

Ich bekomme nie den Leistungswert in Watt ausgegeben, wie das der Vorredner Manos1966 am 04.08.2023 angezeigt bekommen hat.

Bitte Um Hilfe, wie man das hinbekommt.

Und warum gibt Powerfox nicht einfach die Daten lokal frei.
Würde alles um so viel ereleichtern

Danke für sachdienliche Hilfe


![Annotation 2025-03-17 174228](https://github.com/user-attachments/assets/bc12306b-7569-4eaf-8789-9cc84b3184b1)
![Annotation 2025-03-17 174334](https://github.com/user-attachments/assets/627eff06-8c25-4d22-8fe6-80f32f99169b)

**Reply by Billys67** [reply]
Hi, ich habe es nicht mehr verfolgt und dazwischen auf Shelly 3em umgestellt. Es funktioniert mit open battery viel Zuverlässiger als der Poweropti.

**Reply by BigT71** [reply]
Hallo,

bei mir leider das gleiche. Grundsätzlich aber muss es möglich sein. Die powerfox Integration im Home Assistent zeigt den Verbrauch und die beiden Zählerstände. Das hat mich auf eine abenteuerliche Idee gebracht. Ist es möglich den Verbrauch per Home Assistent abzufragen? 

Der läuft eh und zeigt mir den Verbrauch an.

Grüße Thomas

**Reply by Opeloli** [reply]
Hallo BigT71,

klar geht das. Über Homeassistant geht fast alles.
Problematisch dabei:
- Auch HA muss erst einlesen, rechnen, ausgeben weshalb wir schon über nicht unerhebliche Verzögerungen sprechen, die das System zum Schwingen anregen können.

-Wenn sich aus irgendeinem Grund HA aufhängt oder seine Rechenplattform neu bootet, funktioniert die Regelung nicht mehr richtig, was im Idealfall nur verschenken Strom bedeutet, im schlimmsten Fall aber zu Bauteilschädigungen oder schlimmerem führt.
Die Eleganteste und einfachste Lösung ist die direkte Kommunikation des Strommessers mit dem Trucki T2SG.

Und technisch ist das auch trivial, sofern die Daten im lokalen Netzwerk liegen und auslesbar sind.

**Reply by BigT71** [reply]
Ich weiss, die Idee finde ich selbst schlecht. Als Mieter habe ich das Problem das ich eine Lösung möglichst ohne Netzteil brauche. Da ist der Powerfox in der ebz Variante genau das richtige. Solange sich niemand an dem Kabel mit der Antenne stört. Lokal lässt sich der Powerfox nicht mehr auslesen. Sie würden an einer Lösung arbeiten, blah blah blah.

Ich wollte es mit der Daten per Home Assistent Lösung abgreifen testen. Der API Zugriff würde mir genügen ... aber klappt nicht 😞 

Grüße
